### PR TITLE
Policies are periodically retrieved by data aggregator

### PR DESCRIPTION
**Pull request by qibobo**
[#127058171]

### CODE REVIEW

File: src/dataaggregator/aggregator/policyPoller.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+	logger     lager.Logger
+	interval   int
+	database   db.DB
+	tick       *time.Ticker
+	doneChan   chan bool
+	PolicyMap  map[string]*policy.PolicyJson
+	TriggerMap map[string]*policy.Trigger
+}
+
+func NewPolicyPoller(logger lager.Logger, interval int, database db.DB) *PolicyPoller {
```
Comment (fraenkel) [inline src/dataaggregator/aggregator/policyPoller.go]: interval should be a time.Duration

File: src/dataaggregator/aggregator/policyPoller.go [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+	return &PolicyPoller{
+		logger:     logger,
```
Comment (fraenkel) [inline src/dataaggregator/aggregator/policyPoller.go]: I would have created a new logger.WithSession("policy-poller") so we know what this is logging.

File: src/dataaggregator/aggregator/policyPoller.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+			return
+		case <-p.tick.C:
+		}
+	}
+}
+
+func (p *PolicyPoller) retrievePolicy() {
+	policies, err := p.database.RetrievePolicies()
+	if err != nil {
+		p.logger.Error("refresh-app", err)
```
Comment (fraenkel) [inline src/dataaggregator/aggregator/policyPoller.go]: Shouldn't this be retrieve-policies?

File: src/dataaggregator/aggregator/policyPoller.go [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+		return
+	}
+	p.logger.Info("policy count", lager.Data{"count": len(policies)})
+	for _, policyRow := range policies {
+		appId := policyRow.AppId
+		storedPolicyJson, ok := p.PolicyMap[policyRow.AppId]
+		if ok {
```
Comment (fraenkel) [inline src/dataaggregator/aggregator/policyPoller.go]: This is highly redundant. Just determine if things need to be updated and have one block.

File: src/dataaggregator/aggregator/policyPoller_test.go [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+			poller.Stop()
+		})
+		Context("when the poller is started", func() {
+			BeforeEach(func() {
+				database.RetrievePoliciesStub = func() ([]*PolicyJson, error) {
+					return []*PolicyJson{&PolicyJson{AppId: testAppId1, PolicyStr: policyStr1}}, nil
+				}
+			})
+			It("should retrieve policies for every interval", func() {
+				Eventually(database.RetrievePoliciesCallCount, 2500*time.Millisecond, 500*time.Millisecond).Should(Equal(2))
```
Comment (fraenkel) [inline src/dataaggregator/aggregator/policyPoller_test.go]: Once you switch to use a Duration, this can be simplified to use the defaults values.

File: src/dataaggregator/aggregator/policyPoller_test.go [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+			})
+		})
+		Context("when there is no policy", func() {
+			BeforeEach(func() {
+				database.RetrievePoliciesStub = func() ([]*PolicyJson, error) {
+					return []*PolicyJson{}, nil
+				}
+			})
+			It("no trigger generated", func() {
+				Consistently(func() int { return len(poller.PolicyMap) }, 2500*time.Millisecond, 500*time.Millisecond).Should(Equal(0))
```
Comment (fraenkel) [inline src/dataaggregator/aggregator/policyPoller_test.go]: I am going to let this slide for now, but I currently don't see how the poller is going to be used.
I would have expected that something is triggered after a poll which isn't the case.

File: src/dataaggregator/config/config.go [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+type LoggingConfig struct {
+	Level string `yaml:"level"`
+}
+
+var defaultLoggingConfig = LoggingConfig{
+	Level: DefaultLoggingLevel,
+}
+
+type PolicyPollerConfig struct {
+	PollInterval int `yaml:"poll_interval"`
```
Comment (fraenkel) [inline src/dataaggregator/config/config.go]: time.Duration

File: src/dataaggregator/config/config.go [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+var defaultPolicyPollerConfig = PolicyPollerConfig{
+	PollInterval: DefaultPollInterval,
+}
+
+type DbConfig struct {
+	PolicyDbUrl string `yaml:"policy_db_url"`
+}
+
+type Config struct {
```
Comment (fraenkel) [inline src/dataaggregator/config/config.go]: Do we need it to be nested? If we only have 1 config item per, just pull them up. Do the grouping once we have more than a few.

File: src/dataaggregator/config/config.go [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+	Poller  PolicyPollerConfig `yaml:"poller"`
+}
+
+func LoadConfig(reader io.Reader) (*Config, error) {
+	conf := &Config{
+		Logging: defaultLoggingConfig,
+		Poller:  defaultPolicyPollerConfig,
+	}
+
+	decoder := candiedyaml.NewDecoder(reader)
```
Comment (fraenkel) [inline src/dataaggregator/config/config.go]: Please use goyaml. Long story.

File: src/dataaggregator/config/config.go [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+
+	return conf, nil
+}
+
+func (c *Config) Validate() error {
+
+	if c.Db.PolicyDbUrl == "" {
+		return fmt.Errorf("Configuration error: Policy DB url is empty")
+	}
+
```
Comment (fraenkel) [inline src/dataaggregator/config/config.go]: What should happen if the interval is 0?

File: src/dataaggregator/config/config_test.go [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+  poll_interval: "NotInteger"
+`)
+			})
+
+			It("should error", func() {
+				Expect(err).To(MatchError(MatchRegexp("Invalid integer:.*")))
+			})
+		})
+
+		Context("with valid yaml", func() {
```
Comment (fraenkel) [inline src/dataaggregator/config/config_test.go]: This should be first... It is hard to find the good case.

File: src/dataaggregator/db/db.go [hunk 12]
```diff
@@ -0,0 +1,10 @@
+package db
+
+import (
+	"dataaggregator/policy"
+)
+
+type DB interface {
```
Comment (fraenkel) [inline src/dataaggregator/db/db.go]: If this is from metricscollector, then we should pull this out of both to src/db and share across them.

File: src/dataaggregator/policy/policy.go [hunk 13]
```diff
@@ -0,0 +1,53 @@
+package policy
+
+import (
+	"encoding/json"
+)
+
+type IPolicy interface {
```
Comment (fraenkel) [inline src/dataaggregator/policy/policy.go]: We are not doing IPolicy. Call it Policy or ScalingPolicy.

File: src/dataaggregator/policy/policy.go [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+			return true
+		} else {
+			return false
+		}
+	} else {
+		return false
+	}
+}
+func (p *PolicyJson) GetTrigger() *Trigger {
+	appId := p.AppId
```
Comment (fraenkel) [inline src/dataaggregator/policy/policy.go]: Why do we have this variable?

File: src/dataaggregator/policy/policy_test.go [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+		Context("when only one of p1 and p2 are is nil", func() {
+			BeforeEach(func() {
+				p1 = nil
+				p2 = &PolicyJson{AppId: testAppId, PolicyStr: policyStr}
+			})
+			It("should return false", func() {
+				Expect(p1.Equals(p2)).To(Equal(false))
+			})
+		})
+		Context("when p1 and p2 are both not nil", func() {
```
Comment (fraenkel) [inline src/dataaggregator/policy/policy_test.go]: Remove this context.

File: src/dataaggregator/policy/policy_test.go [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+			Context("when the AppId and PolicyStr are both the same", func() {
```
Comment (fraenkel) [inline src/dataaggregator/policy/policy_test.go]: Split this into Context("when the appIds are the same")

File: src/dataaggregator/policy/policy_test.go [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+		})
+	})
+	Context("GetTriggerFromPolicy", func() {
+
+		BeforeEach(func() {
+			policyJson = &PolicyJson{AppId: testAppId, PolicyStr: policyStr}
+			trigger = policyJson.GetTrigger()
+		})
+		It("should return a trigger", func() {
+			Expect(trigger.AppId).To(Equal(testAppId))
```
Comment (fraenkel) [inline src/dataaggregator/policy/policy_test.go]: You could just do a Equals(&Trigger {
....
})
rather that each line which could overlook new fields.

File: src/dataaggregator/aggregator/policyPoller_test.go (hunk 6 above)
Comment (qibobo) [inline src/dataaggregator/aggregator/policyPoller_test.go]: For this story "Policies are periodically retrieved by data aggregator", we just retrieve policies and parse triggers. We will go to next step to aggregate the memory datas.

Now when I use ginkgo -r -race , it will show some warning msg as below. Seems Ben has the same issue. Could you give some advice?
Warning msg:
WARNING: DATA RACE
Read by goroutine 10:
  dataaggregator/aggregator/fakes.(*FakeDB).RetrievePolicies()
      /Users/qiyang/dev/openjavaspace/app-autoscaler/src/dataaggregator/aggregator/fakes/fake_db.go:33 +0x22d
  dataaggregator/aggregator.(*PolicyPoller).retrievePolicies()
      /Users/qiyang/dev/openjavaspace/app-autoscaler/src/dataaggregator/aggregator/policyPoller.go:59 +0x7c
  dataaggregator/aggregator.(*PolicyPoller).startPolicyRetrieve()
      /Users/qiyang/dev/openjavaspace/app-autoscaler/src/dataaggregator/aggregator/pol

Comment (fraenkel) [inline src/dataaggregator/aggregator/policyPoller_test.go]: You need to examine the stacks between the WRITE and READ. In most cases it is more a matter of adding sync blocks. I had you add -race because I could see the issue in how you are testing which is why I asked how the Poller is going to be used.

I would personally break this apart into 2 pieces.
1. A poller that calls some operation on a periodic basis that can be stopped.
2. The ability to retrieve policies
3. There is something that is going to be called after retrieving policies which is passed the new set of policies

It becomes easier to test and by having these additional entry points, you can do the testing on the right goroutines.

File: src/dataaggregator/aggregator/policyPoller_test.go [hunk 18]
```diff
@@ -51,46 +53,47 @@ var _ = Describe("PolicyPoller", func() {
...
 	})
 	Context("Stop", func() {
-		JustBeforeEach(func() {
-			poller.Stop()
-		})
-
 		BeforeEach(func() {
 			poller.Start()
+			poller.Stop()
+			clock.Increment(time.Duration(5 * TestPolicyPollerInterval * time.Second))
```
Comment (fraenkel) [inline src/dataaggregator/aggregator/policyPoller_test.go]: This should be in the test since you are first waiting to see a 1.

File: src/dataaggregator/config/config.go [hunk 19]
```diff
@@ -2,65 +2,44 @@ package config
...
+	"gopkg.in/yaml.v2"
 	"strings"
-
-	"github.com/cloudfoundry-incubator/candiedyaml"
+	"time"
 )
 
 const (
 	DefaultLoggingLevel = "info"
 	DefaultPollInterval = 30
```
Comment (fraenkel) [inline src/dataaggregator/config/config.go]: This should be a duration too

File: src/dataaggregator/config/config.go [hunk 20]
```diff
@@ -2,65 +2,44 @@ package config
...
+	PollInterval time.Duration `yaml:"poll_interval"`
 }
 
-func LoadConfig(reader io.Reader) (*Config, error) {
+func LoadConfig(bytes []byte) (*Config, error) {
 	conf := &Config{
-		Logging: defaultLoggingConfig,
-		Poller:  defaultPolicyPollerConfig,
+		LogLevel:     DefaultLoggingLevel,
+		PollInterval: DefaultPollInterval,
```
Comment (fraenkel) [inline src/dataaggregator/config/config.go]: This will be 30ns instead of 30s given your current default.

**Comment by fraenkel** [comment]
Running with ginkgo -r -p -race I received multiple DATA RACE issues.

**Comment by fraenkel** [comment]
Seems I may have confused you on how this would work.
There shouldn't be a GetTriggers func, it is actually the reverse and closer to what you had. The func is what is called after the new Triggers are built and then passed along to the func for further processing.
There should be no need for any locks since the data is only ever created and passed on the same Go routine.

File: src/dataaggregator/aggregator/policyPoller.go [hunk 21]
```diff
@@ -43,7 +45,12 @@ func (p *PolicyPoller) Stop() {
 }
 func (p *PolicyPoller) startPolicyRetrieve() {
 	for {
-		p.retrievePolicies()
+		policies, error := p.retrievePolicies()
```
Comment (fraenkel) [inline src/dataaggregator/aggregator/policyPoller.go]: Don't name the variable error since that is a type.

File: src/dataaggregator/aggregator/policyPoller.go [hunk 22]
```diff
@@ -52,12 +59,21 @@ func (p *PolicyPoller) startPolicyRetrieve() {
 	}
 }
 
-func (p *PolicyPoller) retrievePolicies() {
+func (p *PolicyPoller) retrievePolicies() ([]*policy.PolicyJson, error) {
 	policies, err := p.database.RetrievePolicies()
 	if err != nil {
 		p.logger.Error("refresh-app", err)
```
Comment (fraenkel) [inline src/dataaggregator/aggregator/policyPoller.go]: Wrong long message

File: src/dataaggregator/aggregator/policyPoller.go [hunk 23]
```diff
@@ -43,7 +45,12 @@ func (p *PolicyPoller) Stop() {
 }
 func (p *PolicyPoller) startPolicyRetrieve() {
 	for {
-		p.retrievePolicies()
+		policies, error := p.retrievePolicies()
+		if error != nil {
+			continue
```
Comment (fraenkel) [inline src/dataaggregator/aggregator/policyPoller.go]: Where is the test for this path?

File: src/dataaggregator/aggregator/policyPoller_test.go [hunk 24]
```diff
@@ -74,48 +66,49 @@ var _ = Describe("PolicyPoller", func() {
...
 	})
 	Context("Stop", func() {
 		BeforeEach(func() {
+			poller = NewPolicyPoller(logger, clock, TestPolicyPollerInterval, database, consumer)
 			poller.Start()
 			poller.Stop()
 		})
 
 		It("stops the polling", func() {
-			clock.Increment(5 * TestPolicyPollerInterval * time.Second)
```
Comment (fraenkel) [inline src/dataaggregator/aggregator/policyPoller_test.go]: Why did you remove this? Without adjusting time, you haven't proven that Stop really works.

**Comment by fraenkel** [comment]
Looks good.
Just squash these into 1 commit.

**Comment by qibobo** [comment]
One commit now. Thank you very much!

**Comment by fraenkel** [comment]
Merged
